src/entities/fractions/price.rs

```diff
@@ -89,15 +89,18 @@ where
     }
 
     /// Return the amount of quote currency corresponding to a given amount of the base currency
-    pub fn quote(&self, currency_amount: CurrencyAmount<TBase>) -> CurrencyAmount<TQuote> {
-        assert!(
-            currency_amount
-                .meta
-                .currency
-                .equals(&self.meta.base_currency),
-            "TOKEN"
-        );
-        let fraction = self.as_fraction() * currency_amount.as_fraction();
+    pub fn quote(
+        &self,
+        currency_amount: CurrencyAmount<TBase>,
+    ) -> Result<CurrencyAmount<TQuote>, Error> {
+        if !currency_amount
+            .meta
+            .currency
+            .equals(&self.meta.base_currency)
+        {
+            return Err(Error::NotEqual("the comparison are not equal".to_owned()));
+        }
+        let fraction = (self.as_fraction()? * currency_amount.as_fraction()?)?.clone();
```

Comment:
i actually wanted to do this, but somehow, they are all linked way to the top, the easiest and possible way to solve this that i saw was to remove this ```impl<T: CurrencyTrait> CurrencyAmount<T> {
    // Constructor method for creating a new currency amount
    fn new(
        currency: T,
        numerator: impl Into<BigInt>,
        denominator: impl Into<BigInt>,
    ) -> Result<Self, Error> {
        let numerator = numerator.into();
        let denominator = denominator.into();
        // Ensure the amount does not exceed MAX_UINT256
        if !numerator.div_floor(&denominator).le(&MAX_UINT256) {
            return Err(Error::MaxUint);
        }
        let exponent = currency.decimals();
        FractionTrait::new(
            numerator,
            denominator,
            CurrencyMeta {
                currency,
                decimal_scale: BigUint::from(10u64).pow(exponent as u32),
            },
        )
    }```

Reply:
Not sure we're on the same page. What I suggested was making `as_fraction` and `mul` return `Self` instead of `Result`. Rationale:

1. `FractionLike::new` only errors if the denominator is 0.
2. Any existing `FractionLike` must have nonzero denominator.
3. Converting any `FractionLike` to a base `Fraction` is impossible to error, which means we can safely unwrap `FractionTrait::new` in `as_fraction` and return `Self`.
4. When adding, subtracting or multiplying two `FractionLike`, the denominator also cannot be zero. Therefore we should return `Self` instead of `Result` for these ops for the sake of readability, dev experience and performance.

Reply:
agree

Reply:
the fraction::new should not return an errror instead, in my opinion it should panic

Reply:
It would make the code a lot cleaner. For example in Solidity, dividing by 0 is a panic.